### PR TITLE
Add NeuTra reparam example

### PR DESCRIPTION
**Pull request by neerajprad**
This adds a NeuTra reparametrizer example on a banana shaped distribution. See [neutra.pdf](https://github.com/pyro-ppl/pyro/files/4150898/neutra.pdf), following discussion in #2294. 

Regarding the API, here are a few things that might need further discussion (separately from this PR):
 - I have changed the `AutoDiagonalNormal` and `AutoMultivariateNormal` guides to use a `TransformedDistribution` so that we can use it with `NeuTraReparam`. 
   - Should we expose `transform` and `base_dist` be available as properties for these classes directly?
   - I think these two are good enough, but I'm not sure if `LowRankMultivariateNormal` should also be written as a `TransformedDistribution`. 
 - Currently, we require users to set `guide.requires_grad_(False)` before passing this to `NeuTraReparam`, since we don't want guide parameters to be updated while running HMC. We can probably do this inside `NeuTraRepam` itself by default (if the guide is a `PyroModule` instance), but I'm not sure if changing the `requires_grad` flag for parameters in the param store is something that might be a surprising side effect for users. Alternative suggestions are welcome. cc. @martinjankowiak

### CODE REVIEW

File: examples/neutra.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    def sample(self, sample_shape=()):
+        u = self.mvn.sample(sample_shape)
+        u0, u1 = u[..., 0], u[..., 1]
+        a, b = self.a, self.b
+        x = a * u0
+        y = (u1 / a) + b * (u0 ** 2 + a ** 2)
+        return torch.stack([x, y], -1)
+
+    def log_prob(self, x):
```
Comment (martinjankowiak) [inline examples/neutra.py]: this is a properly normalized probability?

Comment (neerajprad) [reply examples/neutra.py]: It should be since the determinant of the jacobian is 1.

File: pyro/infer/autoguide/guides.py [hunk 2]
```diff
@@ -589,7 +591,9 @@ def get_posterior(self, *args, **kwargs):
         """
         Returns a diagonal Normal posterior distribution.
         """
-        return dist.Normal(self.loc, self.scale).to_event(1)
+        transform = dist.transforms.AffineTransform(self.loc, self.scale)
+        return dist.TransformedDistribution(dist.Normal(0., 1.).expand([self.latent_dim]).to_event(1),
+                                            transform)
```
Comment (fritzo) [inline pyro/infer/autoguide/guides.py]: 1. Please measure memory and time overhead introduced by this refactoring. This is a heavily used autoguide, and should be kept simple and fast. I am even inclined to introduce a new method `get_posterior_transformed_dist()` or something so as not to obfuscate this code.

2. Please take care with device (the proposed change disregards non-default device):
    ```py
    base_dist = dist.Normal(torch.zeros_like(self.loc), torch.ones_like(self.scale))
    return dist.TransformedDistribution(base_dist.to_event(1), transform)
    ```

File: pyro/infer/autoguide/guides.py [hunk 3]
```diff
@@ -545,7 +545,9 @@ def get_posterior(self, *args, **kwargs):
         """
         Returns a MultivariateNormal posterior distribution.
         """
-        return dist.MultivariateNormal(self.loc, scale_tril=self.scale_tril)
+        transform = dist.transforms.LowerCholeskyAffine(self.loc, scale_tril=self.scale_tril)
```
Comment (fritzo) [inline pyro/infer/autoguide/guides.py]: Please measure memory and time overhead introduced by this refactoring. This is a heavily used autoguide, and should be kept simple and fast. I am even inclined to introduce a new method `.get_posterior_transformed_dist()` or something so as not to obfuscate this code.

**Comment by fritzo** [comment]
> we require users to set `guide.requires_grad_(False)`

Interesting use case. From our 1.0 perspective, `PyroModule`s are the source of truth and the param store is only a legacy convenience for optimization-based inference. So it seems fine to support `nn.Module` methods like `.requires_grad_()`. I guess we could even add a method `.freeze()` or something that converted all `nn.Parameter`s and `PyroParam`s to registered buffers; this would improve performance for constrained parameters since `transform_to` would be called only once during `.freeze()` rather than each time parameters were accessed.

> Should we expose `transform` and `base_dist` ... as properties for these classes directly?

Good idea, I find this cleaner than the proposed refactoring of `.get_posterior()`. Maybe `.get_base_dist()` and `.get_transform()` which together create a `TransformedDistribution` that is equivalent to the simple `dist.Normal(self.loc, self.scale).to_event(1)` returned by `.get_posterior()`?

```py
dist.TransformedDistribution(..., dist.transforms.AffineTransform(...))
```
This is neat! I wonder if it's worth adding a multivariate de-centering reparameterizer, say a `LocScaleTrilReparam` analogous to the existing `LocScaleReparam`. Or we could implement this as a `TransformedDistributionReparam` that re-wrote `MultivariateNormal` as `TransformedDistribution(Normal(zeros,ones), ...)` as you have done in this PR, and maybe registered rewritings of other distributions as `TransformedDistribution` (we could then apply `TransformReparam` on the result).

File: pyro/infer/autoguide/guides.py (hunk 3 above)
Comment (neerajprad) [reply pyro/infer/autoguide/guides.py]: I think we should add `get_base_dist` and `get_transform` methods, and keep the existing `get_posterior` methods as is (we can add a default `get_posterior` though as you suggest in `AutoContinuous`).

Comment (fehiepsi) [reply pyro/infer/autoguide/guides.py]: Agreed! We should make separate `get_base_dist` and `get_transform` methods.

File: pyro/infer/autoguide/guides.py (hunk 2 above)
Comment (neerajprad) [reply pyro/infer/autoguide/guides.py]: Reverted as discussed.

**Comment by neerajprad** [comment]
> From our 1.0 perspective, PyroModules are the source of truth and the param store is only a legacy convenience for optimization-based inference. So it seems fine to support nn.Module methods like .requires_grad_().

Note that we already have `.requires_grad_` available to us via `nn.Module` which we are using in the examples and tests to freeze the guide parameters after training (though that will still run `transform_to` as you mention). @martinjankowiak and I were discussing if we should set `guide.requires_grad_(False)` by default inside `NeuTraReparam` rather than expect users to do so (current behavior). I think doing it inside `NeuTra` is fine too, except that it will have the side effect that the guide passed into `__init__` may have its parameters' `requires_grad` flag toggled to `False`.

> I guess we could even add a method .freeze() or something that converted all nn.Parameters and PyroParams to registered buffers

That will be a nice optimization. We could also modify the `nn.Module.train` and `nn.Module.eval` methods to freeze / unfreeze. I think we could potentially get bigger savings if we are able to jit compile neutra reparameterized models (#2292), but I'll keep this in mind since this is a more general change. 

> I wonder if it's worth adding a multivariate de-centering reparameterizer, say a LocScaleTrilReparam analogous to the existing LocScaleReparam. 

That will be nice, we have all the pieces in place already!

File: examples/neutra.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    fit_guide(guide, args)
+    with pyro.plate('N', args.num_samples):
+        guide_samples = guide()['x'].detach().cpu().numpy()
+
+    ax6.contourf(x1, x2, p, cmap='OrRd')
+    ax6.set(xlabel='x0', ylabel='x1', xlim=xlim, ylim=ylim,
+            title='Posterior \n(BNAF autoguide)')
+    sns.kdeplot(guide_samples[:, 0], guide_samples[:, 1], ax=ax6)
+
+    # 3(b). Draw samples using NeuTra HMC
```
Comment (fritzo) [inline examples/neutra.py]: nit: `4(b)`?

File: pyro/infer/autoguide/guides.py [hunk 5]
```diff
@@ -721,10 +721,10 @@ class AutoNormalizingFlow(AutoContinuous):
     """
 
     def __init__(self, model, init_transform_fn):
+        super().__init__(model, init_loc_fn=init_to_feasible)
```
Comment (fritzo) [inline pyro/infer/autoguide/guides.py]: Do I understand correctly that this failed because `self.transform` was a `TransformModule` and thus required `nn.Module.__init__()` to be called before setting that attribute?

File: examples/neutra.py (hunk 4 above)
Comment (neerajprad) [reply examples/neutra.py]: Thanks for the catch, just fixed.

File: pyro/infer/autoguide/guides.py (hunk 5 above)
Comment (neerajprad) [reply pyro/infer/autoguide/guides.py]: `transform` is just set to None in `__init__` to begin with, but this was needed to call `PyroModule.__init__` before setting `self._prototype_tensor`.

**Comment by neerajprad** [comment]
Thanks for the review and design discussion, @fritzo, @martinjankowiak!
